### ISSUE TITLE
Pravets: fix or remove the printer card support for converting from Latin to Kir chars

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#916]
Spun off from #912.

Basically the function `Pravets::ConvertToPrinterChar(BYTE value)` is broken after switching from ANSI to UTF-8 (initially broken 2 years ago from PR #604).

The problem is there's no test (unit or regression) for this, so any fix will need to be validated.

Perhaps better to just remove this feature?

**Comment by audetto** [comment]
Question 1: is this a feature of the actual computer or the printer attached to it? did they send a font or just a byte to be converted by the printer firmware.
So, who decides that 65 is A? The computer sending a dot matrix or the printer's firmware?

Question 2: what does a real printer do for _non printable characters_?

**Comment by audetto** [comment]
Something like this https://github.com/AppleWin/AppleWin/pull/917?

**Comment by tomcw** [comment]
Thanks - that looks good.
I'll accept the PR, but leave this issue open, as it'd be good to add a regression test.

---

As for your questions:

>Question 1: is this a feature of the actual computer or the printer attached to it? did they send a font or just a byte to be converted by the printer firmware.
>So, who decides that 65 is A? The computer sending a dot matrix or the printer's firmware?

Not the Pravets computer... it'll be the printer interface card's firmware or the printer itself.

>Question 2: what does a real printer do for _non printable characters_?

Depends on the printer interface card and/or the printer.
EG. some will just accept 7-bit values, and _non printable characters_ will do special operations, eg. 0x0D/CR=carriage return, 0x0C/FF=form feed (ie. move to the next sheet of paper).
And some will also accept 8-bit values, and maybe the characters with bit7=1 will print special characters.

**Comment by sicklittlemonkey** [comment]
Hi guys, great work tidying this up. It did make the core print routine a
bit messy.

Characters printed are determined by the printer. Most of the parallel or
serial interface cards only handle flow control and escaping etc. The
character sets are in the printer.

Cheers,
Nick.

**Comment by audetto** [comment]
> It did make the core print routine a bit messy.

I agree. If there was a data type to represent a single code point, it would probably be better.
Could ``wchar_t`` help?

**Comment by tomcw** [comment]
Perhaps you could programmatically convert the UTF-8 strings back to ANSI, eg. see the hoops I had to jump through to the other way (ANSI -> unicode -> UTF-8), here:

https://github.com/AppleWin/AppleWin/blob/735317ae17ade0c12ba738f4e1fdc5f7d59129ea/source/YamlHelper.cpp#L456

**Comment by tomcw** [comment]
Here what I mean by ANSI -> unicode -> UTF-8: #920

This keeps all the original logic, and just does an upfront string conversion in the ctor.

**Comment by sicklittlemonkey** [comment]
I managed to get a printer manual from the Apple II Bulgaria Facebook group. Download here:
https://drive.google.com/file/d/1SBkQQ0Wvh66P7dMO7hyhmXfM9U4BFqhS/view?usp=sharing

The important page is the character set in this image:
![PravetzPrinterASCIITable](https://user-images.githubusercontent.com/4393005/106142496-489e9e00-61bd-11eb-8f2e-e4937e7becb5.jpg)

**Comment by audetto** [comment]
Which does not seem related with https://github.com/AppleWin/AppleWin/blob/master/source/Pravets.cpp#L213
Unless there has already been a codepage mistake somewhere.

The current code replaces lower cases with accents, while the picture above replaces with Cyrillic character, which makes much more sense.
And it would match what gets printed on screen.

Could it be that the *correct* conversion makes the text mode screen and the printer file very similar?

**Comment by audetto** [comment]
Similar but not the same, the characters in the range 123-126 are not the same as printed on screen for any of the clones.

**Comment by tomcw** [comment]
You can see from page-16 of the printer manual, then the "Cyrillic text on screen"(*) matches the printed Cyrillic text.

(*) Actually this is printed text too (eg. with a LIST command), but if it wasn't what was on screen, then it'd be very hard to write this BASIC code!
